_posts/2025_01_08_exploiting_capabilities.markdown

```diff
@@ -0,0 +1,73 @@
+And of course there are a lot of memory safety benefits that come for free from that, but the real value is unlocked when you start to manipulate them directly to capture intent.
+Let me illustrate with a simple example.
+
+The publish method in the MQTT library in [cheriot-network](https://github.com/CHERIoT-Platform/network-stack.git) takes pointers for both the topic and the payload:
+
+```c++
+int __cheri_compartment("MQTT") mqtt_publish(Timeout    *t,
+                                             SObj        mqttHandle,
+                                             uint8_t     qos,
+                                             const char *topic,
+                                             size_t      topicLength,
+                                             const void *payload,
+                                             size_t      payloadLength,
+                                             bool        retain = false);
+```
+
+and has the following comment about their scope:
+
+_Both the topic and payload buffers must remain valid during the execution of this function._
+_If the caller frees them during the execution of this function, the publish may leak application data to the broker through the topic and payload._
+
+Ok, so it seems clear that I only need to make sure that they remain valid and unchanged until the call returns, but I was hitting a bug that seemed to suggest they might be being used outside of that by another thread in the network stack.
+_It turns out I was wrong about that BTW, but anyway I'm a suspicious kind of guy when it comes to trusting external modules._
+A quick look though the code quickly persuaded me that a) I'm not very good at reading other peoples code and b) it was probably not the issue but I still wanted to rule it out.
+
+Luckily by thinking of these in terms of capabilities rather than pointers I can directly implement my intent rather that trust what the library is telling me.
+Specifically I can derive new capabilities that are read only and can't be captured.
+So I can be guaranteed that at the end of the call nothing in the network stack can have modified the message or saved a pointer to in.
+And for good measure I can also limit the bounds so that I don't have to trust topicLength and payloadLength will be used properly.
+This is what the resulting code looks like in my demo:
+
+```c++
+// Publish a string to the status topic
+void publish(SObj mqtt, std::string topic, void *status, bool retain)
+{
+	// Use capabilities with only Load
+	// permission so we can be sure the MQTT stack
+	// doesn't capture a pointer to our buffer
+	Timeout           t{MS_TO_TICKS(5000)};
+	
+	CHERI::Capability roTopic{topic.data()};
+	roTopic.permissions() &= {CHERI::Permission::Load};
+	roTopic.bounds() = topic.size();
+
+	CHERI::Capability roStatus{status};
+	roStatus.permissions() &= {CHERI::Permission::Load};
```

Comment:
Shouldn't you set the bounds on `roStatus` too?

Reply:
In the demo this is taken from its not needed because that's already a capability with the right bounds.  I could pass in a size_t to make the example more complete, but rather that add that I just put a footnote explaining this - hope that's OK.

Reply:
Can you add a comment saying that the bounds are already correct then?

Reply:
Done